File: android/app/src/main/java/deckers/thibault/aves/model/provider/ImageProviderFactory.java
package deckers.thibault.aves.model.provider;

import android.content.ContentResolver;
import android.net.Uri;
import android.provider.MediaStore;

import androidx.annotation.NonNull;

public class ImageProviderFactory {
    public static ImageProvider getProvider(@NonNull Uri uri) {
        String scheme = uri.getScheme();

        if (ContentResolver.SCHEME_CONTENT.equalsIgnoreCase(scheme)) {
            // a URI's authority is [userinfo@]host[:port]
            // but we only want the host when comparing to Media Store's "authority"
            if (MediaStore.AUTHORITY.equalsIgnoreCase(uri.getHost())) {
                return new MediaStoreImageProvider();
            }
            return new ContentImageProvider();
        }

        if (ContentResolver.SCHEME_FILE.equalsIgnoreCase(scheme)) {
            return new FileImageProvider();
        }

        return null;
    }
}

File: android/app/src/main/kotlin/deckers/thibault/aves/model/provider/ImageProviderFactory.kt
package deckers.thibault.aves.model.provider

import android.content.ContentResolver
import android.content.Context
import android.net.Uri
import deckers.thibault.aves.utils.StorageUtils
import java.util.Locale

object ImageProviderFactory {
    fun getProvider(context: Context, uri: Uri): ImageProvider? {
        return when (uri.scheme?.lowercase(Locale.ROOT)) {
            ContentResolver.SCHEME_CONTENT -> {
                if (StorageUtils.isMediaStoreContentUri(uri)) {
                    MediaStoreImageProvider()
                } else if (AvesEmbeddedMediaProvider.provides(context, uri)) {
                    AvesEmbeddedMediaProvider()
                } else {
                    UnknownContentProvider()
                }
            }

            ContentResolver.SCHEME_FILE -> FileImageProvider()
            else -> null
        }
    }
}